core/src/main/java/com/yahoo/oak/EntryArray.java

```diff
@@ -0,0 +1,369 @@
+         */
+        VALID;
+
+        /**
+         * We consider an entry to be valid if it was inserted and not deleted.
+         *
+         * @return is the value valid
+         */
+        boolean isValid() {
+            return this.ordinal() >= EntryState.VALID.ordinal();
+        }
+    }
+
+
+    /********************************************************************************************/
+    /*---------------- Methods for setting and getting specific entry's field ------------------*/
+
+    /**
+     * Converts external entry-index to internal array index.
+     * <p>
+     * We use the following terminology:
+     *  - longIdx for the long's index inside the entries array (referred as a set of longs)
+     *  - entryIdx for the index of entry array (referred as a set of entries)
+     *
+     * @param entryIdx external entry-index
+     * @return the internal array index
+     */
+    private int entryIdx2LongIdx(int entryIdx) {
+        return entryIdx * fields;
+    }
+
+
+    /**
+     * Returns the number of entries allocated and not deleted for this EntrySet.
+     * Although, in case EntrySet is used as an array, nextFreeIndex is can be used to calculate
+     * number of entries, additional variable is used to support OakHash
+     */
+    int getNumOfEntries() {
+        return numOfEntries.get();
+    }
+
+    /**
+     * Atomically reads long field of the entries array.
+     */
+    protected long getEntryFieldLong(int entryIndex, int entryOffset) {
+        return entries[entryIdx2LongIdx(entryIndex) + entryOffset];
+    }
+
+    protected void setEntryFieldLong(int entryIndex, int entryOffset, long value) {
+        entries[entryIdx2LongIdx(entryIndex) + entryOffset] = value;
+    }
+
+    /**
+     * Performs CAS of given long field of the entries longs array,
+     * that should be associated with some entry.
+     * CAS from 'expectedLongValue' to 'newLongValue' for field at specified offset
+     *
+     */
+    protected boolean casEntryFieldLong(int entryIndex, int entryOffset, long expectedLongValue,
+                                      long newLongValue) {
+        int index = Unsafe.ARRAY_LONG_BASE_OFFSET +
+                (entryIdx2LongIdx(entryIndex) + entryOffset) * Unsafe.ARRAY_LONG_INDEX_SCALE;
+        return UnsafeUtils.UNSAFE.compareAndSwapLong(entries, index, expectedLongValue, newLongValue);
+    }
+
+
+    /********************************************************************************************/
+    /*--------------- Methods for setting and getting specific key and/or value ----------------*/
+
+    /**
+     * Atomically reads the key reference from the entry (given by entry index "ei")
+     */
+    protected long getKeyReference(int ei) {
+        return getEntryFieldLong(ei, KEY_REF_OFFSET);
+    }
+
+    /**
+     * Atomically writes the key reference to the entry (given by entry index "ei")
+     */
+    protected void setKeyReference(int ei, long value) {
```

Comment:
Probably to add casKeyReference, as it would be needed for OakHash

Reply:
IMO, it is best to add methods when needed. Otherwise, you wouldn't have a code path to this method and you wouldn't be able to test it.

Reply:
OK, skipping this comment